### DISCUSSION TITLE
why SWTICHTAG + PERTAG does not switch to tag

### DISCUSSION THREAD

**Discussion by yayuwang0624** [bakkeby/dwm-flexipatch#355] [Q&A] [answered]
if `SWTICHTAG_PATCH` enabled, `void applyrules(Client *c)` will do 

```c
pertagview(&((Arg) { .ui = newtagset }));
arrange(c->mon);
```
instead of 
```c
view(&((Arg) { .ui = newtagset }));
```

but it actually doesn't switch to tag automatically. 

Is it a design or a bug?

**Comment by bakkeby** [comment]
If in your client rules you say that a window (let's say firefox) open on tag 8, then it will only change the view to tag 8 when that window opens if the rule includes `.switchtag = 1`, or have a 1 in the switchtag column if you are using traditional columns for configuring your rules. 

The general principle of the switchtag patch is identical to:
https://dwm.suckless.org/patches/switchtotag/

**Reply by yayuwang0624** [reply]
That is what I expected but the fact is that the view doesn't change if I set .switchtag to 1 or 3.

dwm switches to tag 8 when I open firefox for example only if I disable pertag or change the source to 

```c
pertagview(&((Arg) { .ui = newtagset }));
arrange(c->mon);
view(&((Arg) { .ui = newtagset }));
```

**Reply by bakkeby** [reply]
Do you also set the tags?

**Reply by yayuwang0624** [reply]
for sure. can you confirm that `pertagview` will also set the view to `newtagset` like `view`?

**Reply by bakkeby** [reply, marked as answer]
I have had a chance to test this now and yes I think it is a bug. I don't remember the reasoning for using pertagview here, but I have probably missed this due to also using the focusonnetactive patch.

I'll change this to just view for now.

In general it is bad design to change the view in the applyrules function (it is a hack basically) - that should really just apply rules to a client and not act upon them. I think there are some edge cases where this changes the view when it shouldn't. For example let's say that the user is on tag 3 and starts firefox in a terminal window. The browser has a client rule saying that it should be placed on tag 8 and that the view should change to that when it starts, so when rules are applied the view is changed to tag 8. But, the user has the swallow patch as well and firefox is swallowed by the terminal (or firefox swallows the terminal, whichever way you look at it) and instead it is shown on tag 3 where the terminal lives. Now the user has been moved to tag 8 where there are no clients.
